Cesium.CodeGen/Ir/Expressions/SizeOfOperatorExpression.cs

```diff
@@ -0,0 +1,54 @@
+using Cesium.Ast;
+using Cesium.CodeGen.Contexts;
+using Cesium.CodeGen.Extensions;
+using Cesium.CodeGen.Ir.Declarations;
+using Cesium.CodeGen.Ir.Types;
+
+namespace Cesium.CodeGen.Ir.Expressions;
+
+internal class SizeOfOperatorExpression : IExpression
+{
+    private readonly IType _type;
+
+    public SizeOfOperatorExpression(IType Type)
+    {
+        _type = Type;
+    }
+
+    public SizeOfOperatorExpression(Ast.SizeOfOperatorExpression sizeOfOperator)
+    {
+        var (specifiers, abstractDeclarator) = sizeOfOperator.TypeName;
+        _type = (specifiers, abstractDeclarator) switch
+        {
+            ({ }, { }) => LocalDeclarationInfo.Of(specifiers, abstractDeclarator).Type,
+            ({ }, null) => LocalDeclarationInfo.Of(specifiers, (Declarator?)null).Type
+        };
+    }
+
+    public IExpression Lower(IDeclarationScope scope)
+    {
+        if (_type is NamedType namedType)
+        {
+            var typeResolved = new IdentifierExpression(namedType.TypeName).Resolve(scope).GetValueType();
```

Comment:
Currently, this definition, and perhaps even parsing for this, is incorrect.

We want to support both `sizeof(variable)` and `sizeof(type)`, right?

But so far the parser only supports `sizeof(type_name)`, and thus it _accidentally_ supports _variables_ that _look like type names_, which cannot be right 😅 

This leads to further problems, one of which is now demonstrated in a failed `typedef` test I pushed into the branch. In my opinion, it looks like you are trying to save the incorrect definition by resolving a `NamedType` as a variable name.

This is of course move in the right direction, but let's work a little bit more on the definition of this: perhaps let's introduce two different `sizeof` variants for types and variables? (and then accept just an `identifier` and see how it resolves in the lower/emit time)

Note I'm okay to disable the (currently) failing tests and merge this while opening a follow-up issue on this, but probably I'd like to disable the support for _variables_ completely beforehand. I am open for discussion, though. If you are very eager to merge this as is, I'm ready to negotiate somewhat :)

Reply:
At first I had two different expressions for this but somehow it worked having only one of them as `sizeof(type_name)`, sure I will split both scenarios and I will update this pull request along the other issue mentioned above.

Thank you for the feedback!